src/pokemon.rs

```diff
@@ -15,8 +19,176 @@ pub struct Pokemon {
-    let pokemon: Vec<Pokemon> = serde_json::from_str(pokemon_json_str)?;
-    Ok(pokemon)
+impl Pokemon {
+    /// Get all forms of the pokemon except those in the exclude vector
+    pub fn get_filtered_forms(&self, exclude: &[&str]) -> Vec<String> {
+        self.forms
+            .iter()
+            .filter(|f| !exclude.contains(&f.as_str()))
+            .cloned()
+            .collect()
+    }
+}
+
+trait GenerationParser {
+    fn parse_generations(&self) -> Result<Vec<u8>, Error>;
+}
+
+impl GenerationParser for &str {
+    fn parse_generations(&self) -> Result<Vec<u8>, Error> {
+        let gens: Vec<u8> = if let Some((start_gen, end_gen)) = self.split_once('-') {
+            let start_gen = start_gen.parse::<u8>();
+            let end_gen = end_gen.parse::<u8>();
+            match (start_gen, end_gen) {
+                (Ok(s), Ok(e)) => (s..=e).collect(),
+                _ => return Err(Error::InvalidGeneration(self.to_string())),
+            }
+        } else {
+            self.split(',')
+                .map(|gen| gen.parse::<u8>())
+                .collect::<Result<Vec<_>, _>>()
+                .map_err(|_| Error::InvalidGeneration(self.to_string()))?
+        };
+
+        if gens.is_empty() {
+            return Err(Error::InvalidGeneration(self.to_string()));
+        }
+
+        Ok(gens)
+    }
+}
+
+pub struct PokemonDatabase {
+    pokemon: Vec<Pokemon>,
+    config: Config,
+}
+
+impl PokemonDatabase {
+    pub fn load(pokemon_db: &EmbeddedFile, config: Config) -> Result<Self, Error> {
+        let pokemon_json_str =
+            str::from_utf8(&pokemon_db.data).expect("Invalid UTF-8 in pokemon db");
+        let pokemon: Vec<Pokemon> = serde_json::from_str(pokemon_json_str)?;
+        Ok(Self { pokemon, config })
+    }
+
+    /// Filter pokemon by generation
+    pub fn filter_by_generation(&self, generations: &str) -> Result<Vec<&Pokemon>, Error> {
+        let gens = generations.parse_generations()?;
+
+        let pokemon: Vec<&Pokemon> = self
+            .pokemon
+            .iter()
+            .filter(|p| gens.contains(&p.gen))
+            .collect();
+
+        Ok(pokemon)
+    }
+
+    /// Returns a vector of all pokemon
+    pub fn get_all(&self) -> &Vec<Pokemon> {
+        &self.pokemon
+    }
+
+    /// Prints the names of all pokemon from the given generations into stdout
+    pub fn list_pokemon_names(&self, generations: &str) -> Result<(), Error> {
+        let filtered_pokemon = self.filter_by_generation(generations)?;
+        filtered_pokemon.iter().for_each(|p| println!("{}", p.slug));
+        Ok(())
+    }
+
+    /// Prints the pokemon with the given name into stdout
+    pub fn show_pokemon_by_name(&self, name: &Name) -> Result<(), Error> {
+        let pokemon = self
+            .get_all()
+            .iter()
+            .find(|p| p.slug == name.name)
+            .ok_or_else(|| Error::InvalidPokemon(name.name.clone()))?;
```

Comment:
while we're at it, we could validate the provided form also, instead of panicking on invalid forms, eg:
```diff
+        if name.form != "regular" && !pokemon.forms.contains(&name.form) {
+            return Err(Error::InvalidPokemonForm(
+                name.name.clone(),
+                name.form.clone(),
+            ));
+        }
```

```diff
@@ -11,6 +11,9 @@ pub enum Error {
     #[error("Invalid pokemon `{0}`")]
     InvalidPokemon(String),
 
+    #[error("Invalid form `{1}` for pokemon `{0}`")]
+    InvalidPokemonForm(String, String),
+

```